ext/strings_test.go

```diff
@@ -329,3 +335,673 @@ func TestStringsWithExtension(t *testing.T) {
+			dynArgs: map[string]any{
+				"dynList": []any{6, 4.2, "a string"},
+			},
+			expectedOutput: `dyntype list: [6, 4.200000, "a string"]`,
+		},
+		{
+			name:       "dyntype support for maps",
+			format:     "dyntype map: %s",
+			formatArgs: `dynMap`,
+			dynArgs: map[string]any{
+				"dynMap": map[any]any{
+					"strKey": "x",
+					true:     42,
+					int64(6): mustParseDuration("7m2s"),
+				},
+			},
+			expectedOutput: `dyntype map: {"strKey":"x", 6:duration("422s"), true:42}`,
+		},
+		{
+			name:       "unrecognized formatting clause",
+			format:     "%a",
+			formatArgs: "1",
+			err:        "could not parse formatting clause: unrecognized formatting clause \"a\"",
+		},
+		{
+			name:       "out of bounds arg index",
+			format:     "%d %d %d",
+			formatArgs: "0, 1",
+			err:        "index 2 out of range",
+		},
+		{
+			name:       "string substitution is not allowed with binary clause",
+			format:     "string is %b",
+			formatArgs: `"abc"`,
+			err:        "error during formatting: only integers and bools can be formatted as binary, was given string",
+		},
+		{
+			name:       "duration substitution not allowed with decimal clause",
+			format:     "%d",
+			formatArgs: `duration("30m2s")`,
+			err:        "error during formatting: decimal clause can only be used on integers, was given google.protobuf.Duration",
+		},
+		{
+			name:       "string substitution not allowed with octal clause",
+			format:     "octal: %o",
+			formatArgs: `"a string"`,
+			err:        "error during formatting: octal clause can only be used on integers, was given string",
+		},
+		{
+			name:       "double substitution not allowed with hex clause",
+			format:     "double is %x",
+			formatArgs: "0.5",
+			err:        "error during formatting: only integers, byte buffers, and strings can be formatted as hex, was given double",
+		},
+		{
+			name:       "uppercase not allowed for scientific clause",
+			format:     "double is %E",
+			formatArgs: "0.5",
+			err:        `could not parse formatting clause: unrecognized formatting clause "E"`,
+		},
+		{
+			name:       "object not allowed",
+			format:     "object is %s",
+			formatArgs: `ext.TestAllTypes{PbVal: test.TestAllTypes{}}`,
+			err:        "error during formatting: string clause can only be used on strings, bools, bytes, ints, doubles, maps, lists, durations, and timestamps, was given ext.TestAllTypes",
+		},
+		{
+			name:       "type() not allowed",
+			format:     "type is %s",
+			formatArgs: `type("test string")`,
+			err:        "error during formatting: string clause can only be used on strings, bools, bytes, ints, doubles, maps, lists, durations, and timestamps, was given type",
+		},
+		{
+			name:       "null not allowed for %s",
+			format:     "null: %s",
+			formatArgs: "null",
+			err:        "error during formatting: string clause can only be used on strings, bools, bytes, ints, doubles, maps, lists, durations, and timestamps, was given null_type",
```

Comment:
Allowing null for %s seems _very_ useful, particularly for optional fields.

Reply:
Right, `null` can be used with `%s` now. (Documented this as well.)